src/scaladoc/scala/tools/nsc/doc/base/CommentFactoryBase.scala

```diff
@@ -544,6 +568,340 @@ trait CommentFactoryBase { this: MemberLookupBase =>
+      * cell       ::= nPipe content
+      * nPipe      ::= '|'+
+      * content    ::= { '\n'+ } { inline } { content }
+      * }}}
+      */
+    def table(): Block = {
+
+      /* Helpers */
+
+      def peek(tag: String): Unit = {
+        val peek: String = buffer.substring(offset)
+        val limit = 60
+        val limitedPeek = peek.substring(0, limit min peek.length)
+        println(s"peek: $tag: '$limitedPeek'")
+      }
+
+      def nextIsCellStart = check(TableCellStart)
+
+      /* Accumulated state */
+
+      var header: Option[Row] = None
+
+      val rows = mutable.ListBuffer.empty[Row]
+
+      val cells = mutable.ListBuffer.empty[Cell]
+
+      def finalizeCells(): Unit = {
+        if (cells.nonEmpty) {
+          rows += Row(cells.toList)
+        }
+        cells.clear()
+      }
+
+      def finalizeHeaderCells(): Unit = {
+        if (cells.nonEmpty) {
+          if (header.isDefined) {
+            reportError(pos, "more than one table header")
+          } else {
+            header = Some(Row(cells.toList))
+          }
+        }
+        cells.clear()
+      }
+
+      def checkAny(terminators: List[String]) = terminators.exists(check)
+
+      def isEndOfText = char == endOfText
+
+      def isNewline = char == endOfLine
+
+      def isNewlineAnyTableMarks = checkAny(TableAllMarksOnNextLine)
+
+      def skipNewline() = jump(endOfLine)
+
+      /*
+       * Make progress but not by jumping all newlines which would prevent the inline method from
+       * leaving the input buffer in a state that can be handled by the cases in this method.
+       */
+      def skipInsignificantNewlines() = while (isNewline && !isNewlineAnyTableMarks) {
+        skipNewline()
+      }
+
+      def contentNonEmpty(content: Inline) = content != Text("")
+
+      /**
+        * @param nextIsStartMark True if the next char is a cell mark prefix and not any non-cell mark.
+        * @param cellStartMark   The char the cell start mark is based on
+        * @param finalizeRow     Function to invoke when the row has been fully parsed
+        */
+      def parseCells(nextIsStartMark: => Boolean, cellStartMark: Char, finalizeRow: () => Unit): Unit = {
+        /* The first sequence of cellStartMark characters defines the markdown for new cells. */
+        def parseStartMark() = {
+          val startCount = repeatJump(cellStartMark)
+          cellStartMark.toString * startCount
+        }
+
+        /* startMark is the only mark not requiring a newline first */
+        def makeInlineTerminators(startMark: String) = startMark :: TableAllMarksOnNextLine
+
+        val startPos = offset
+
+        val startMark = parseStartMark()
+
+        if (startMark.isEmpty) {
+          peek("Expected startMark")
+          sys.error("Precondition violated: Expected startMark.")
+        }
+
+        val inlineTerminators = makeInlineTerminators(startMark)
+
+        val content = Paragraph(inline(isInlineEnd = checkAny(inlineTerminators)))
```

Comment:
Is it true that all terminators are single chars? Then this could be made more efficient by just checking the buffer's current `char` instead of calling `check` many times. (We can do this in a follow-up PR). A first simple improvement would be to test in `check` if the `chars` string has length 1, and avoid calling `jump` in that case.

Reply:
I'm doing the clean up now so will look at this.